007-OFEP-flag-change-events.md

```diff
@@ -0,0 +1,66 @@
+
+Per-flag design:
+
+```ts
+const client = OpenFeature.getClient();
+
+// subscribe to changes in a string flag called "hex-color", and evaluate it when it's updated
+client.addHandler("hex-color", async (key) => {
+  console.log(
+    `Got update for ${key}, new value is ${await client.getStringValue(
+      key,
+      "000000"
+    )}`
+  );
+});
+```
+
+General configuration change design:
+
+```ts
+const client = OpenFeature.getClient();
+
+// subscribe to changes in the configuration in general, and evaluate flagd it when it's updated
+// this might be easier to implement across vendors, since many don't allow subscription to a particular flag, but general configuration changes
+client.on("configuration-change", async (changeMetadata) => {
+  console.log(
+    `Got update for configuration, new value for hex-color is ${await client.getStringValue(
+      key,
+      "000000"
+    )}`
+  );
+});
+```
+
+## Benefits
+
+Ability to use event-based flag evaluation paradigms.
+
+## Caveats
+
+- Not all providers can reasonably implement this... some SDKs don't support subscriptions or configuration change notifications, for example. Should these providers simply never fire events?
+- Implementation in providers will likely be more divergent than imperative evaluation (based on demo work below).
+- Re-evaluating a flag after the handler fires sometimes means another round-trip, but most of the time in this case, things are locally cached - really dependant on SDK implementation. This is the pattern used by some vendors though, so it's not unprecedented.
+- Some SDKs can't identify individual flags have changed, so all registered handlers must fire (see cloudbees provider demo). This might not be such a big deal since they are likely no-ops.
+- We may need a way to "shutdown" providers, cancelling all handlers.
+- We may need a way to "move" handlers from one provider to another if the provider is changed (recreate them in the new provider).
+- We should probably consider other events, like "SDK-READY" as well, which would fit well into design 2, above.
```

Comment:
I think an event like SDK-READY would make alot of sense.   The current evaluation spec lets us provide a default value, which could be returned during an application startup before the SDK is ready, but there is no way for the consumer to know they got the default vs what they may have wanted.  Being able to listen of SDK-READY effectively lets them handle this, without blocking an application from starting up.